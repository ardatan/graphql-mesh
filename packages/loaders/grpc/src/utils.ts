import { existsSync } from "fs";
import type { SchemaComposer } from "graphql-compose";
import { isAbsolute, join } from "path";
import type { INamespace, Root } from "protobufjs";
import { getGraphQLScalarForGrpc, isGrpcScalar } from "./scalars.js";
import lodashHas from 'lodash.has';

export function addIncludePathResolver(root: Root, includePaths: string[]): void {
  const originalResolvePath = root.resolvePath;
  root.resolvePath = (origin: string, target: string) => {
    if (isAbsolute(target)) {
      return target;
    }
    for (const directory of includePaths) {
      const fullPath: string = join(directory, target);
      if (existsSync(fullPath)) {
        return fullPath;
      }
    }
    const path = originalResolvePath(origin, target);
    if (path === null) {
      console.warn(`${target} not found in any of the include paths ${includePaths}`);
    }
    return path;
  };
}

export function getTypeName(
  schemaComposer: SchemaComposer,
  pathWithName: string[] | undefined,
  isInput: boolean,
) {
  if (pathWithName?.length) {
    const baseTypeName = pathWithName.filter(Boolean).join('__');
    if (isGrpcScalar(baseTypeName)) {
      return getGraphQLScalarForGrpc(baseTypeName);
    }
    if (schemaComposer.isEnumType(baseTypeName)) {
      return baseTypeName;
    }
    return isInput ? baseTypeName + '_Input' : baseTypeName;
  }
  return 'Void';
}

export function walkToFindTypePath(
  rootJson: INamespace,
  pathWithName: string[],
  baseTypePath: string[],
) {
  const currentWalkingPath = [...pathWithName];
  while (!lodashHas(rootJson.nested, currentWalkingPath.concat(baseTypePath).join('.nested.'))) {
    if (!currentWalkingPath.length) {
      break;
    }
    currentWalkingPath.pop();
  }
  return currentWalkingPath.concat(baseTypePath);
}
