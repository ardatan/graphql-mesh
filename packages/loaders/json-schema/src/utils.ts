import { OperationTypeNode } from 'graphql';
import { JSONSchemaOperationConfig, JSONSchemaPubSubOperationConfig, HTTPMethod } from './types';

export function isPubSubOperationConfig(
  operationConfig: JSONSchemaOperationConfig
): operationConfig is JSONSchemaPubSubOperationConfig {
  return 'pubsubTopic' in operationConfig;
}

export function getOperationMetadata(operationConfig: JSONSchemaOperationConfig) {
  let httpMethod: HTTPMethod;
  let operationType: OperationTypeNode;
  let rootTypeName: 'Query' | 'Mutation' | 'Subscription';
  if (isPubSubOperationConfig(operationConfig)) {
    httpMethod = null;
    operationType = 'subscription' as OperationTypeNode;
    rootTypeName = 'Subscription';
  } else {
    httpMethod = operationConfig.method;
    // Fix compability with Mesh handler
    operationType = operationConfig.type.toLowerCase() as OperationTypeNode;
    if (!httpMethod) {
      if (operationType === 'mutation') {
        httpMethod = 'POST';
      } else {
        httpMethod = 'GET';
      }
    }
    if (!rootTypeName) {
      if (httpMethod === 'GET') {
        rootTypeName = 'Query';
      }
    }
    rootTypeName = operationType === 'query' ? 'Query' : 'Mutation';
  }
  return {
    httpMethod,
    operationType,
    rootTypeName,
    fieldName: operationConfig.field,
  };
}
