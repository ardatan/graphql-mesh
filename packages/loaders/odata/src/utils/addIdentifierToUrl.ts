export function addIdentifierToUrl(
  url: URL,
  identifierFieldName: string,
  identifierFieldTypeRef: string,
  args: any,
) {
  url.href += `/${args[identifierFieldName]}/`;
}
