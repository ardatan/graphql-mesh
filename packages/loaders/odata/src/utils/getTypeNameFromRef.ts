import { pascalCase } from 'pascal-case';

interface GetNamespaceFromTypeRefOpts {
  namespaces: Set<string>;
  typeRef: string;
}

function getNamespaceFromTypeRef({ typeRef, namespaces }: GetNamespaceFromTypeRefOpts) {
  let namespace = '';
  namespaces?.forEach(el => {
    if (
      typeRef.startsWith(el) &&
      el.length > namespace.length && // It can be deeper namespace
      !typeRef.replace(el + '.', '').includes('.') // Typename cannot have `.`
    ) {
      namespace = el;
    }
  });
  return namespace;
}

const SCALARS = new Map<string, string>([
  ['Edm.Binary', 'String'],
  ['Edm.Stream', 'String'],
  ['Edm.String', 'String'],
  ['Edm.Int16', 'Int'],
  ['Edm.Byte', 'Byte'],
  ['Edm.Int32', 'Int'],
  ['Edm.Int64', 'BigInt'],
  ['Edm.Double', 'Float'],
  ['Edm.Boolean', 'Boolean'],
  ['Edm.Guid', 'GUID'],
  ['Edm.DateTimeOffset', 'DateTime'],
  ['Edm.Date', 'Date'],
  ['Edm.TimeOfDay', 'String'],
  ['Edm.Single', 'Float'],
  ['Edm.Duration', 'ISO8601Duration'],
  ['Edm.Decimal', 'Float'],
  ['Edm.SByte', 'Byte'],
  ['Edm.GeographyPoint', 'String'],
]);

export function getTypeNameFromRef({
  typeRef,
  isInput,
  isRequired,
  aliasNamespaceMap,
  multipleSchemas,
  namespaces,
  isEnumType,
}: {
  typeRef: string;
  isInput: boolean;
  isRequired: boolean;
  aliasNamespaceMap: Map<string, string>;
  multipleSchemas: boolean;
  namespaces: Set<string>;
  isEnumType(typeName: string): boolean;
}) {
  const typeRefArr = typeRef.split('Collection(');
  const arrayDepth = typeRefArr.length;
  let actualTypeRef = typeRefArr.join('').split(')').join('');
  const typeNamespace = getNamespaceFromTypeRef({
    typeRef: actualTypeRef,
    namespaces,
  });
  if (aliasNamespaceMap.has(typeNamespace)) {
    const alias = aliasNamespaceMap.get(typeNamespace);
    actualTypeRef = actualTypeRef.replace(typeNamespace, alias);
  }
  const actualTypeRefArr = actualTypeRef.split('.');
  const typeName = multipleSchemas
    ? pascalCase(actualTypeRefArr.join('_'))
    : actualTypeRefArr[actualTypeRefArr.length - 1];
  let realTypeName = typeName;
  if (SCALARS.has(actualTypeRef)) {
    realTypeName = SCALARS.get(actualTypeRef);
  } else if (isEnumType(typeName)) {
    realTypeName = typeName;
  } else if (isInput) {
    realTypeName += 'Input';
  }
  const fakeEmptyArr = new Array(arrayDepth);
  realTypeName = fakeEmptyArr.join('[') + realTypeName + fakeEmptyArr.join(']');
  if (isRequired) {
    realTypeName += '!';
  }
  return realTypeName;
}
