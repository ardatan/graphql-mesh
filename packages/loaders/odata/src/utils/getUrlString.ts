export function getUrlString(url: URL) {
  return decodeURIComponent(url.toString()).split('+').join(' ');
}
