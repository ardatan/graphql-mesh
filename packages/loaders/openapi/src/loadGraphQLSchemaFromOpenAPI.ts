import { loadGraphQLSchemaFromJSONSchemas } from '@omnigraph/json-schema';
import { OpenAPILoaderOptions } from '.';
import { getJSONSchemaOptionsFromOpenAPIOptions } from './getJSONSchemaOptionsFromOpenAPIOptions';

/**
 * Creates a local GraphQLSchema instance from a OpenAPI Document.
 * Everytime this function is called, the OpenAPI file and its dependencies will be resolved on runtime.
 * If you want to avoid this, use `createBundle` function to create a bundle once and save it to a storage
 * then load it with `loadGraphQLSchemaFromBundle`.
 */
export async function loadGraphQLSchemaFromOpenAPI(name: string, options: OpenAPILoaderOptions) {
  const extraJSONSchemaOptions = await getJSONSchemaOptionsFromOpenAPIOptions(options);
  return loadGraphQLSchemaFromJSONSchemas(name, {
    ...options,
    ...extraJSONSchemaOptions,
  });
}
