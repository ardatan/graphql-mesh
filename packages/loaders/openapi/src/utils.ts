import { camelCase } from 'change-case';

export function getFieldNameFromPath(path: string, method: string, responseTypeSchemaRef: string) {
  // Replace identifiers with "by"
  path = path.split('{').join('by_').split('}').join('');

  const [actualPartsStr, allQueryPartsStr] = path.split('?');

  const actualParts = actualPartsStr.split('/').filter(Boolean);

  let fieldNameWithoutMethod = actualParts.join('_');

  // If path doesn't give any field name without identifiers, we can use the return type with HTTP Method name
  if ((!fieldNameWithoutMethod || fieldNameWithoutMethod.startsWith('by')) && responseTypeSchemaRef) {
    const refArr = responseTypeSchemaRef.split('/');
    // lowercase looks better in the schema
    const prefix = camelCase(refArr[refArr.length - 1]);
    if (fieldNameWithoutMethod) {
      fieldNameWithoutMethod = prefix + '_' + fieldNameWithoutMethod;
    } else {
      fieldNameWithoutMethod = prefix;
    }
  }

  if (allQueryPartsStr) {
    const queryParts = allQueryPartsStr.split('&');
    for (const queryPart of queryParts) {
      const [queryName] = queryPart.split('=');
      fieldNameWithoutMethod += '_' + 'by' + '_' + queryName;
    }
  }

  // get_ doesn't look good in field names
  if (method === 'get') {
    return fieldNameWithoutMethod;
  }
  return method + '_' + fieldNameWithoutMethod;
}
