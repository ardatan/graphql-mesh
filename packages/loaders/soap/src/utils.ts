import { X2jOptions } from 'fast-xml-parser';

export interface SoapAnnotations {
  elementName: string;
  bindingNamespace: string;
  endpoint: string;
}

export const PARSE_XML_OPTIONS: Partial<X2jOptions> = {
  attributeNamePrefix: '',
  attrNodeName: 'attributes',
  textNodeName: 'innerText',
  ignoreAttributes: false,
  ignoreNameSpace: true,
  arrayMode: true,
  allowBooleanAttributes: true,
};
