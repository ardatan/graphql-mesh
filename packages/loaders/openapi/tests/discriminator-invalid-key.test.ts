import { createRouter, Response } from 'fets';
import { execute, GraphQLSchema, parse } from 'graphql';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { loadGraphQLSchemaFromOpenAPI } from '../src/loadGraphQLSchemaFromOpenAPI.js';

describe('Discriminator with invalid key in the mapping', () => {
  let createdSchema: GraphQLSchema;
  beforeAll(async () => {
    interface Pet {
      name: string;
      petType: string;
    }

    interface Cat extends Pet {
      color: string;
    }

    interface Dog extends Pet {
      age: string;
    }
    const pets: { [id: string]: Cat | Dog } = {
      '1': {
        name: 'Fluffy',
        petType: 'pet-cat',
        color: 'climber',
      },
      '2': {
        name: 'Buddy',
        petType: 'pet-dog',
        age: '23',
      },
    };
    const router = createRouter()
      .route({
        path: '/pets/:id',
        method: 'GET',
        handler(req) {
          const pet = pets[req.params.id];
          if (pet) {
            return Response.json(pet);
          }
          return new Response('Pet not found', {
            status: 404,
          });
        },
      })
      .route({
        path: '/dogs/:id',
        method: 'GET',
        handler(req) {
          const pet = pets[req.params.id];
          if (pet && pet.petType === 'Dog') {
            return Response.json(pet);
          }
          return new Response('Dog not found', {
            status: 404,
          });
        },
      })
      .route({
        path: '/cats/:id',
        method: 'GET',
        handler(req) {
          const pet = pets[req.params.id];
          if (pet && pet.petType === 'Cat') {
            return Response.json(pet);
          }
          return new Response('Cat not found', {
            status: 404,
          });
        },
      });

    createdSchema = await loadGraphQLSchemaFromOpenAPI('test', {
      source: './fixtures/discriminator-invalid-keys.yml',
      endpoint: 'http://localhost',
      cwd: __dirname,
      ignoreErrorResponses: true,
      // @ts-ignore
      fetch: router.fetch,
    });
  });
  it('should generate correct schema', () => {
    expect(printSchemaWithDirectives(createdSchema)).toMatchSnapshot('discriminator-invalid-key');
  });
  it('should handle discriminator mapping', async () => {
    const query = /* GraphQL */ `
      query Pet {
        cat: pets_by_id(id: "1") {
          __typename
          name
          petType
          ... on Cat {
            color
          }
        }
        dog: pets_by_id(id: "2") {
          __typename
          name
          petType
          ... on Dog {
            age
          }
        }
      }
    `;
    const result = await execute({
      schema: createdSchema,
      document: parse(query),
      variableValues: {
        petsByIdId: '1',
      },
    });
    expect(result).toEqual({
      data: {
        cat: {
          __typename: 'Cat',
          name: 'Fluffy',
          petType: 'pet-cat',
          color: 'climber',
        },
        dog: {
          __typename: 'Dog',
          name: 'Buddy',
          petType: 'pet-dog',
          age: '23',
        },
      },
    });
  });
});
