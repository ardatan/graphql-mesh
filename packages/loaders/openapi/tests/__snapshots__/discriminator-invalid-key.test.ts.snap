// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discriminator with invalid key in the mapping should generate correct schema: discriminator-invalid-key 1`] = `
"schema @transport(subgraph: "test", kind: "rest", location: "http://localhost") {
  query: Query
}

directive @discriminator(subgraph: String, field: String, mapping: [[String]]) on INTERFACE | UNION

directive @httpOperation(subgraph: String, path: String, operationSpecificHeaders: [[String]], httpMethod: HTTPMethod, isBinary: Boolean, requestBaseBody: ObjMap, queryParamArgMap: ObjMap, queryStringOptionsByParam: ObjMap, jsonApiFields: Boolean, queryStringOptions: ObjMap) on FIELD_DEFINITION

directive @transport(subgraph: String, kind: String, location: String, headers: [[String]], queryStringOptions: ObjMap, queryParams: [[String]]) repeatable on SCHEMA

type Cat implements Pet {
  color: String
  name: String!
  petType: String
}

interface Pet @discriminator(subgraph: "test", field: "petType", mapping: [["pet-cat", "Cat"], ["pet-dog", "Dog"]]) {
  name: String!
  petType: String
}

type Dog implements Pet {
  age: String
  name: String!
  petType: String
}

type Query {
  pets_by_id(id: String!): Pet @httpOperation(subgraph: "test", path: "/pets/{args.id}", operationSpecificHeaders: [["accept", "application/json"]], httpMethod: GET)
}

enum HTTPMethod {
  GET
  HEAD
  POST
  PUT
  DELETE
  CONNECT
  OPTIONS
  TRACE
  PATCH
}

scalar ObjMap"
`;
